packages/codec/lib/compilations/utils.ts

```diff
@@ -138,17 +159,34 @@ export function shimContracts(
     contracts.push(contractObject);
   }
 
+  if (inputSources) {
+    //if input sources was passed, set up the sources object directly :)
+    sources = inputSources.map(
+      ({sourcePath, contents: source, ast, language}) => ({
+        sourcePath,
+        source,
+        ast: <Ast.AstNode>ast,
+        language
+        //we'll omit compiler, as if inputSources was passed, presumably
+        //we're using shimCompilation(), which sets that up separately
+      })
+    );
+  }
+
   //now: check for id overlap with internal sources
-  for (let contract of contracts) {
-    const { generatedSources, deployedGeneratedSources } = contract;
-    for (let index in generatedSources) {
-      if (index in sources) {
-        unreliableSourceOrder = true;
+  //(don't bother if inputSources or files was passed)
+  if (!inputSources && !files) {
+    for (let contract of contracts) {
+      const {generatedSources, deployedGeneratedSources} = contract;
+      for (let index in generatedSources) {
+        if (index in sources) {
+          unreliableSourceOrder = true;
+        }
       }
-    }
-    for (let index in deployedGeneratedSources) {
-      if (index in sources) {
-        unreliableSourceOrder = true;
+      for (let index in deployedGeneratedSources) {
```

Comment:
Oh btw I keep forgetting to let you know that `for (const` works just fine here

Reply:
Oh yeah I've been a bit inconsistent with that; I have `for (const` in a lot of places, but not everywhere.  Well, I won't change that here right now, but can try to remember for the future...

I think maybe what confused me is that when you do a C-style for loop, then you do need to use `let`?  I think?